src/lib.rs

```diff
@@ -1264,6 +1277,61 @@ where
     }
 }
 
+/// A streaming iterator that flattens nested streaming iterators.
+#[derive(Debug)]
+pub struct Flatten<I, J> {
+    iter: I,
+    _inner: PhantomData<J>,
+}
+
+impl<I, J> StreamingIterator for Flatten<I, J>
+where
+    I: StreamingIteratorMut<Item = J>,
+    J: StreamingIterator,
+{
+    type Item = J::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        loop {
+            if let Some(ref mut iter) = self.iter.get_mut() {
+                iter.advance();
+                if !iter.is_done() {
+                    break;
+                }
+            }
+            self.iter.advance();
+            if self.iter.is_done() {
+                break;
+            }
+        }
+    }
+
+    #[inline]
+    fn is_done(&self) -> bool {
+        match self.iter.get() {
+            Some(iter) => iter.is_done(),
+            None => true,
+        }
+    }
+
+    #[inline]
+    fn get(&self) -> Option<&Self::Item> {
+        self.iter.get().and_then(|iter| iter.get())
+    }
+}
```

Comment:
Please also add `fold` for internal iteration -- probably in terms of `self.iter.fold_mut`.